pkg/util/expiring_cache.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+	"time"
+)
+
+// ExpiringCache is a thread-safe "time expiring" cache that
+// automatically removes objects that are not accessed for a
+// configurable delay
+//
+// It is used in various places where we need to cache data for an
+// unknown amount of time, to prevent memory leaks
+//
+// From the consumer's perspective, it behaves similarly to a map
+// KeyType is the type of the object that is used as a key
+// ValueType is the type of the object that is stored
+type ExpiringCache[KeyType comparable, ValueType any] interface {
+	// Get operates identically to retrieving from a map, returning
+	// the value and/or boolean indicating if the value existed in the map
+	//
+	// Multiple callers can receive the same value simultaneously from Get,
+	// it is the caller's responsibility to ensure they are not modified
+	Get(key KeyType) (value *ValueType, ok bool)
+	// Set operates identically to setting a value in a map, adding an entry
+	// or overriding the existing value for a given key
+	Set(key KeyType, value *ValueType)
+}
+
+type timedValue[ValueType any] struct {
+	value *ValueType
+	timer *time.Timer
+}
+
+type expiringCache[KeyType comparable, ValueType any] struct {
+	expirationDelay time.Duration
+	values          map[KeyType]timedValue[ValueType]
+	mutex           sync.Mutex
+}
+
+// NewExpiringCache returns a new ExpiringCache
+// for a given KeyType, ValueType, and expiration delay
+func NewExpiringCache[KeyType comparable, ValueType any](expirationDelay time.Duration) ExpiringCache[KeyType, ValueType] {
```

Comment:
np: If this file is moved to its own separate package (`expiringcache`) as per previous feedback, I'd suggest renaming this to simply `New`: `expiringcache.New()` > `expiringcache.NewExpiringCache()`.